taichi/backends/codegen_opengl.cpp

```diff
@@ -0,0 +1,534 @@
+#include "codegen_opengl.h"
+#include <taichi/platform/opengl/opengl_api.h>
+
+#include <string>
+#include <taichi/ir.h>
+
+TLANG_NAMESPACE_BEGIN
+namespace opengl {
+namespace {
+
+std::string opengl_data_type_name(DataType dt)
+{
+  switch (dt) {
+    case DataType::f32:
+      return "float";
+    case DataType::i32:
+      return "int";
+    case DataType::u32:
+      return "uint";
+    default:
+      TI_NOT_IMPLEMENTED;
+      break;
+  }
+  return "";
+}
+
+std::string opengl_unary_op_type_symbol(UnaryOpType type)
+{
+  switch (type)
+  {
+  case UnaryOpType::neg:
+    return "-";
+  case UnaryOpType::sqrt:
+    return "sqrt";
+  case UnaryOpType::floor:
+    return "floor";
+  case UnaryOpType::ceil:
+    return "ceil";
+  case UnaryOpType::abs:
+    return "abs";
+  case UnaryOpType::sgn:
+    return "sign";
+  case UnaryOpType::sin:
+    return "sin";
+  case UnaryOpType::asin:
+    return "asin";
+  case UnaryOpType::cos:
+    return "cos";
+  case UnaryOpType::acos:
+    return "acos";
+  case UnaryOpType::tan:
+    return "tan";
+  case UnaryOpType::tanh:
+    return "tanh";
+  case UnaryOpType::exp:
+    return "exp";
+  case UnaryOpType::log:
+    return "log";
+  default:
+    TI_NOT_IMPLEMENTED;
+  }
+  return "";
+}
+
+bool is_opengl_binary_op_infix(BinaryOpType type)
+{
+  return !((type == BinaryOpType::min) || (type == BinaryOpType::max) ||
+           (type == BinaryOpType::atan2) || (type == BinaryOpType::pow));
+}
+
+class KernelGen : public IRVisitor
+{
+  Kernel *kernel;
+
+public:
+  KernelGen(Kernel *kernel, std::string kernel_name)
+    : kernel(kernel), kernel_name_(kernel_name),
+      glsl_kernel_prefix_(kernel_name)
+  {
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+private: // {{{
+  std::string kernel_src_code_;
+  std::string indent_;
+  bool is_top_level_{true};
+
+  const SNode *root_snode_;
+  GetRootStmt *root_stmt_;
+  std::string kernel_name_;
+  std::string glsl_kernel_name_;
+  std::string root_snode_type_name_;
+  std::string glsl_kernel_prefix_;
+  int glsl_kernel_count_{0};
+
+  void push_indent()
+  {
+    indent_ += "  ";
+  }
+
+  void pop_indent()
+  {
+    indent_.pop_back();
+    indent_.pop_back();
+  }
+
+  template <typename... Args>
+  void emit(std::string f, Args &&... args)
+  {
+    kernel_src_code_ +=
+        indent_ + fmt::format(f, std::forward<Args>(args)...) + "\n";
+  }
+
+  void generate_header()
+  {
+    emit("#version 430 core");
+    emit("#extension GL_ARB_compute_shader: enable");
+    emit("");
+    emit("layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;");
+    emit("");
+    emit("layout(std430, binding = 0) buffer data");
+    emit("{{");
+    emit("  int _args_[{}];", taichi_max_num_args);
+    emit("  int _data_[];");
+    emit("}};");
```

Comment:
This layout is the `data` we passed from `load_data`, or ` ssbo.bind_data`. Both input & output.
Also try specify `layout(std430, binding = 0) volatile buffer data`.
`std430` means use standard C-style layout, aligns, etc.